src/module_writer/virtual_writer.rs

```diff
@@ -0,0 +1,294 @@
+    /// Returns Ok(Some(..)) if the new source should be included, Ok(None) if
+    /// the new source should not be included (excluded or duplicate).
+    fn get_entry(
+        &mut self,
+        target: PathBuf,
+        source: Option<&Path>,
+    ) -> Result<Option<VacantEntry<'_, PathBuf, ArchiveSource>>> {
+        if let Some(source) = source {
+            if self.exclude(source) {
+                return Ok(None);
+            }
+        }
+
+        if self.exclude(&target) {
+            if !self.target_exclusion_warning_emitted {
+                self.target_exclusion_warning_emitted = true;
+                eprintln!(
+                    "⚠️ Warning: A file was excluded from the archive by the target path in the archive\n\
+                     ⚠️ instead of the source path on the filesystem. This behavior is deprecated and\n\
+                     ⚠️ will be removed in future versions of maturin.",
+                );
+            }
+            debug!("Excluded file {target:?} from archive by target path");
+            return Ok(None);
+        }
+
+        let entry = match self.tracker.entry(target.clone()) {
+            Entry::Vacant(entry) => Some(entry),
+            Entry::Occupied(entry) => {
+                match (entry.get().path(), source) {
+                    (None, None) => {
+                        bail!(
+                            "Generated file {} was already added, can't add it again",
+                            target.display()
+                        );
+                    }
+                    (Some(previous_source), None) => {
+                        bail!(
+                            "File {} was already added from {}, can't overwrite with generated file",
+                            target.display(),
+                            previous_source.display()
+                        )
+                    }
+                    (None, Some(source)) => {
+                        bail!(
+                            "Generated file {} was already added, can't overwrite it with {}",
+                            target.display(),
+                            source.display()
+                        );
+                    }
+                    (Some(previous_source), Some(source)) => {
+                        if is_same_file(source, previous_source).unwrap_or(false) {
+                            // Ignore identical duplicate files
+                            None
+                        } else {
+                            bail!(
+                                "File {} was already added from {}, can't add it from {}",
+                                target.display(),
+                                previous_source.display(),
+                                source.display()
+                            );
+                        }
+                    }
+                }
+            }
+        };
+
+        Ok(entry)
+    }
+
+    pub(crate) fn add_entry(
```

Comment:
Note that `VirtualWriter` has an `add_entry()` function but does not implement `ModuleWriterInternal` intentionally. Because `VirtualWriter` is generic over any `ModuleWriterInternal` this specific setup prevents creating nested `VirtualWriter<VirtualWriter<..>>` situations.